flink-runtime/src/test/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManagerToTaskExecutorHeartbeatSchedulerTest.java

```diff
@@ -0,0 +1,142 @@
+import static org.junit.Assert.assertTrue;
+
+public class ResourceManagerToTaskExecutorHeartbeatSchedulerTest extends TestLogger {
+
+	@Rule
+	public TemporaryFolder tempFolder = new TemporaryFolder();
+
+	private static TestingServer testingServer;
+
+	private static ActorSystem actorSystem;
+	private static ActorSystem actorSystem2;
+	private static AkkaRpcService akkaRpcService;
+	private static AkkaRpcService akkaRpcService2;
+
+	private static final Timeout timeout = new Timeout(10000, TimeUnit.MILLISECONDS);
+
+	@BeforeClass
+	public static void setup() throws Exception {
+		testingServer = new TestingServer();
+
+		actorSystem = AkkaUtils.createDefaultActorSystem();
+		actorSystem2 = AkkaUtils.createDefaultActorSystem();
+
+		akkaRpcService = new AkkaRpcService(actorSystem, timeout);
+		akkaRpcService2 = new AkkaRpcService(actorSystem2, timeout);
+	}
+
+	@AfterClass
+	public static void teardown() throws Exception {
+		akkaRpcService.stopService();
+		akkaRpcService2.stopService();
+
+		actorSystem.shutdown();
+		actorSystem2.shutdown();
+
+		actorSystem.awaitTermination();
+		actorSystem2.awaitTermination();
+
+		testingServer.stop();
+	}
+
+	@Test
+	public void testStart() throws Exception {
+		ExecutorService executorService = new ForkJoinPool();
+
+		Configuration configuration = ZooKeeperTestUtils
+			.createZooKeeperRecoveryModeConfig(
+				testingServer.getConnectString(),
+				tempFolder.getRoot().getPath());
+
+		CuratorFramework client = ZooKeeperUtils.startCuratorFramework(configuration);
+		LeaderElectionService leaderElectionService = ZooKeeperUtils.createLeaderElectionService(client,
+			configuration);
+		ResourceManager resourceManager = new ResourceManager(akkaRpcService, executorService, leaderElectionService);
+
+		HighAvailabilityServices highAvailabilityServices = new NonHaServices(resourceManager.getAddress());
+		ResourceID resourceID = ResourceID.generate();
+		TaskExecutor taskExecutor = new TaskExecutor(akkaRpcService2, highAvailabilityServices, resourceID);
+
+		resourceManager.start();
+		taskExecutor.start();
+
+		ResourceManagerToTaskExecutorHeartbeatScheduler heartbeatScheduler = new ResourceManagerToTaskExecutorHeartbeatScheduler(resourceManager, resourceManager.getLeaderSessionID(), taskExecutor.getSelf(), taskExecutor.getAddress(), resourceID, log);
+		heartbeatScheduler.start();
+		Assert.assertFalse(heartbeatScheduler.isClosed());
+		// if heartbeat scheduler cannot receive heartbeat response from taskExecutor for maxAttempt times, it will close itself.
+		// else it will continue to trigger next heartbeat in the given interval milliseconds
+		FiniteDuration timeout = new FiniteDuration(40, TimeUnit.SECONDS);
+		Deadline deadline = timeout.fromNow();
+
+		while (deadline.hasTimeLeft() && !heartbeatScheduler.isClosed()) {
+			Thread.sleep(100);
+		}
+
+		Assert.assertFalse(heartbeatScheduler.isClosed());
+
+		heartbeatScheduler.close();
+		Assert.assertTrue(heartbeatScheduler.isClosed());
+	}
```

Comment:
I think we should have more tests, testing the different heartbeat cases. 1. receiving a regular heartbeat in time  and checking that the slot report is properly delivered. 2. Irregular heartbeat 3. Loss of heartbeat and check that the resource is properly failed. Etc.